### DISCUSSION TITLE
How long does indexing take?

### DISCUSSION THREAD

**Discussion by bdochosiete** [ordinals/ord#1791] [Q&A] [answered]
I'm using an old Mac with 8g RAM but I'm actually running Ubuntu and using an external SSD to store the bitcoin node. Syncing the whole Blockchain took about 2 days. Indexing ORD wallet looks like it will take about 20 days, it started fast but at around 50% progress it now takes about an hour to go through 1000 blocks, what's the usual processing time for this? And I'm curious about the indexing process, what's happening there? Is it traversing all previous blocks with every new block?

**Comment by ghost** [comment, marked as answer]
> Indexing ORD wallet looks like it will take about 20 days, it started fast but at around 50% progress it now takes about an hour to go through 1000 blocks, what's the usual processing time for this?

This PR could improve things if you can build from source: https://github.com/casey/ord/pull/1759

```

$ rustc --version
$ cargo --version

$ git clone --single-branch https://github.com/andrewtoth/ord.git --branch batch-tx-requests

$ cd ord

$ cargo build

$ cargo run
```

**Reply by apemithrandir** [reply]
> > Indexing ORD wallet looks like it will take about 20 days, it started fast but at around 50% progress it now takes about an hour to go through 1000 blocks, what's the usual processing time for this?
> 
> This PR could improve things if you can build from source: #1759
> 
> ```
> 
> $ rustc --version
> $ cargo --version
> 
> $ git clone --single-branch https://github.com/andrewtoth/ord.git --branch batch-tx-requests
> 
> $ cd ord
> 
> $ cargo build
> 
> $ cargo run
> ```

That helps immensely for regular index. the --index-sats will still run in it's usual time.

**Reply by bdochosiete** [reply]
For a moment I thought it wasn't helping but it did help enormously. Thanks!

**Reply by GaloisField2718** [reply]
How to be sure this repo is safe and it will not take control or attack our computer or our data ?

**Reply by GaloisField2718** [reply]
> > Indexing ORD wallet looks like it will take about 20 days, it started fast but at around 50% progress it now takes about an hour to go through 1000 blocks, what's the usual processing time for this?
> 
> This PR could improve things if you can build from source: #1759
> 
> ```
> 
> $ rustc --version
> $ cargo --version
> 
> $ git clone --single-branch https://github.com/andrewtoth/ord.git --branch batch-tx-requests
> 
> $ cd ord
> 
> $ cargo build
> 
> $ cargo run
> ```

I think the last command is `cargo run wallet balance` and not only `cargo run` because it opens help prompt.

**Reply by bdochosiete** [reply]
Yeah, it's just telling you that that's similar to run ORD, you would have to add subcommands etc...

**Reply by GaloisField2718** [reply]
Just a question 
> That helps immensely for regular index. the --index-sats will still run in it's usual time.

when we are running `ord wallet balance`, we need to add --index-sats ?

**Reply by bdochosiete** [reply]
The way I'm running it is: cargo run index. But since my cookie isn't in the default directory I'm specifying the cookie path. So it actually looks like this: cargo run -- --cookie-file /path/to/.cookie index.

**Reply by bdochosiete** [reply]
I realized it hasn't finish yet, it stopped at block 767465 and it's now going 1 block per minute, which it's about 7 days time to index the whole thing, why does that happen?

**Reply by GaloisField2718** [reply]
> I realized it hasn't finish yet, it stopped at block 767465 and it's now going 1 block per minute, which it's about 7 days time to index the whole thing, why does that happen?

Have you run `bitcoind -txindex` ?

**Reply by GaloisField2718** [reply]
Without `-txindex` flag indeed it isn't working and `ord` tells it at the end

**Reply by bdochosiete** [reply]
Yeah, I'm running bitcoind, I modified my txindex=1 in the Bitcoin Core configuration file so I don't need to specify the -txindex flag. I mean it's running, it just got super slow at the very end.

**Reply by GaloisField2718** [reply]
Okay ^^' weird. No idea I'm still with bitcoind which downloading so I will check after this ^^

**Reply by GaloisField2718** [reply]
And same as previously, killall ord shutdown computer and rerun all ?

**Reply by bdochosiete** [reply]
I've tried that but it will basically start from where it first stopped, so it went from 0 to 767368 and then at some point it made it to 768013, I stopped everything, rebooted and reran commands but it did the same, it started from 0 :S.

**Reply by GaloisField2718** [reply]
Ah yes ^^' me it raises an error near to the end `[indexing blocks] ███████████████████████████████████████████████░ 771440/777567` 
which I don't understand : 
``` 
[2023-02-21T03:37:46Z ERROR ord::index::updater] Couldn't receive txs connection error: Connection reset by peer (os error 54)
error: failed to get transaction for 5c4398575e6d87189d900f6c1ce8bd72cee2ceb6f542a0e21e9dd5c0d7217e61
```

I will put an issue, but if it's related to this branch here could be more efficient.

**Reply by bdochosiete** [reply]
I got to the point in which Bitcoin blocks are being solved faster than my indexing process 😂. I'm gonna drop the external SSD and try again within my computers drive.

**Reply by goddesslens** [reply]
> I got to the point in which Bitcoin blocks are being solved faster than my indexing process 😂. I'm gonna drop the external SSD and try again within my computers drive.

External SSDs are 5 times slower than internal NVMe SSDs. Even with 2000MB/s read write speed it took me two days to index. What's your read write speed ? External SSD USB will make it much more slower.

**Reply by bdochosiete** [reply]
Yeah that was my suspicion. The thing is I wanted to keep everything on an external drive to be able to switch between computers. This old computer has an HDD, since the external SDD is limited to USB 3 speed, it's practically the same speed as the internal Hard drive. I'll give it a go with the internal HDD, if it doesn't work then I'll just move into my personal computer, I just wanted to recycle this old thing for something useful.

**Reply by GaloisField2718** [reply]
> > I got to the point in which Bitcoin blocks are being solved faster than my indexing process 😂. I'm gonna drop the external SSD and try again within my computers drive.
> 
> External SSDs are 5 times slower than internal NVMe SSDs. Even with 2000MB/s read write speed it took me two days to index. What's your read write speed ? External SSD USB will make it much more slower.

Oh no :( It's exactly what I want to do also. I don't want to kill my macbook and on SSD it might less compute and get back some storage ^^'

**Reply by bdochosiete** [reply]
I tried to run it in my internal HDD and it's way slower, the external SSD is definitely a better option. I tried syncing the bitcoin Blockchain and it was so slow, with the external SSD I'm able to catch up to the latest block within 5 minutes (after a day of not syncing), the internal HDD would take a day.

 It's just that last part of the indexing that I'm having issues with, no idea why it gets so slow at the end.

**Reply by GaloisField2718** [reply]
A question do you finished ord indexing ?

**Reply by bdochosiete** [reply]
Negative, you?

**Reply by bdochosiete** [reply]
Got to 768119/777770 and it takes for ever to move to the next one. This isn't related to the speed of my hard drive, this time I used the internal hard drive.

**Reply by GaloisField2718** [reply]
> Negative, you?

Same… each time I finish it it tells me error RPC Connection rather than my bitcoind runs without error… I really don’t understand…

**Reply by GaloisField2718** [reply]
> Got to 768119/777770 and it takes for ever to move to the next one. This isn't related to the speed of my hard drive, this time I used the internal hard drive.

Ah yeah it definitely moves slow down near to the end whatever you are or you’re using….

**Reply by GaloisField2718** [reply]
That's it !!!! I've done ! But it's really not as expected ^^'
I must work enough. Because now I only posted basic txt files and each time my ord wallet had been emptying and I can't look them

**Reply by bdochosiete** [reply]
I finally managed to sync and index everything. I ditched Ubuntu and did everything from scratch on Catalina MacOS. Thankfully, the commit for faster indexing was part of the newest release for ORD so I didn't have to mess much with the setup. Everything runs way smoother now, with Ubuntu I couldn't do anything else than the running bitcoind and ord.

**Reply by GoodbyeMatrix-Github** [reply]
> > Indexing ORD wallet looks like it will take about 20 days, it started fast but at around 50% progress it now takes about an hour to go through 1000 blocks, what's the usual processing time for this?
> 
> This PR could improve things if you can build from source: #1759
> 
> ```
> 
> $ rustc --version
> $ cargo --version
> 
> $ git clone --single-branch https://github.com/andrewtoth/ord.git --branch batch-tx-requests
> 
> $ cd ord
> 
> $ cargo build
> 
> $ cargo run
> ```

Hi everyone!

I am having the same issue while indexing in Ubuntu. I am trying this solution but I am having the following error while executing the command: "git clone --single-branch https://github.com/andrewtoth/ord.git --branch batch-tx-requests":

"Cloning into 'ord'...
warning: Could not find remote branch batch-tx-requests to clone.
fatal: Remote branch batch-tx-requests not found in upstream origin"


I have the following versions installed:

- ord 0.5.1
- rustc 1.67.1 (d5a82bbd2 2023-02-07)
- cargo 1.67.1 (8ecd4f20a 2023-01-10)

Any suggestions?

Thanks in advance!

**Reply by bdochosiete** [reply]
Since the mainline branch already has theses changes included, I would just follow the [Readme.md](https://github.com/casey/ord) instructions to install/build ORD. That worked for me.

**Reply by GoodbyeMatrix-Github** [reply]
> Since the mainline branch already has theses changes included, I would just follow the [Readme.md](https://github.com/casey/ord) instructions to install/build ORD. That worked for me.

It worked with the commands in the Readme.md file!

Thank you very much for your assistance!
